lesson_21/homework_21.py

```diff
@@ -0,0 +1,114 @@
+
+Обов’язково включіть результат роботи — файл hb_test.log в PR.
+"""
+
+
+class HBAnalyzer:
+    """Class to analyze heartbeat logs."""
+
+    def __init__(self, log_file, key, timestamp_prefix,
+                 warn_limit, err_limit, time_format='%H:%M:%S'):
+        """Initialize parameters."""
+        self.log_file = log_file
+        self.key = key
+        self.timestamp_prefix = timestamp_prefix
+        self.warn_limit = warn_limit
+        self.err_limit = err_limit
+        self.time_format = time_format
+        self.logger = self._configure_logger()
+
+    def _configure_logger(self):
+        """Configure logger."""
+        logger = logging.getLogger(__name__)
+        logger.setLevel(logging.INFO)
+        handler = logging.FileHandler('hb_test.log', mode='a')
+        formatter = logging.Formatter('%(asctime)s : %(levelname)s - '
+                                      '%(message)s', '%Y-%m-%d %H:%M:%S')
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+        return logger
+
+    def _read_lines(self):
+        """Read lines from the log file."""
+        try:
+            with open(self.log_file, 'rt') as file:
+                lines = file.readlines()
+                if not lines:
+                    self.logger.warning(f'The log file {self.log_file}'
+                                        f' is empty.')
+                return lines
+        except FileNotFoundError:
+            self.logger.error(f'File not found: {self.log_file}')
+            return []
+
+    def _extract_timestamp(self, line):
+        """Extract timestamp from a log line."""
+        start_idx = line.find(self.
+                              timestamp_prefix) + len(self.timestamp_prefix)
+        timestamp_str = line[start_idx:start_idx + 8]
+        try:
+            return datetime.strptime(timestamp_str, self.time_format)
+        except (ValueError, IndexError):
+            self.logger.error(f'Invalid timestamp '
+                              f'format in line: {line.strip()}')
+            return None
+
+    def log_analyzer(self):
+        """Analyze the log file for heartbeat issues."""
+        lines = self._read_lines()
+        filtered_lines = [line for line in lines if self.key in
+                          line and self.timestamp_prefix in line]
+
+        if not filtered_lines:
+            self.logger.warning(f'No lines found with the key {self.key}')
+            return
+
+        timestamps = [self._extract_timestamp(line) for line in
+                      filtered_lines if self._extract_timestamp(line)]
+        for i in range(len(timestamps) - 1):
+            self._log_difference(timestamps[i], timestamps[i + 1])
+
+    def _log_difference(self, time_a, time_b):
+        """Log the difference between two timestamps."""
+        time_diff = abs((time_b - time_a).total_seconds())
+        if time_diff >= self.err_limit:
+            self.logger.error(f'Heartbeat delay {time_diff}'
```

Comment:
The most important info is interval `time_a till  time_b` - I mean exact place , exact timestamp in logs where we see error or warning diff. So better to add this info to output logging too